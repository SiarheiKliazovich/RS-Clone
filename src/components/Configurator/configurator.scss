@import "../../sass/variables";
@import "../../sass/mixins";


.configurator {
  max-width: 1200px;
  margin: 0 auto;
}

.active-link {
  display: block;
  padding: 3px 0;
  color: $color-white1;
  text-decoration: none;
  font-size: $font-size-xxlg;
  font-family: $font-family-base;
  background-color: transparent;

  &::before {
    background: $color-white1;
    bottom: 39px;
    content: "";
    height: 2px;
    left: 8px;
    position: absolute;
    width: 100%;
    transition: all .3s ease-in-out;
    transform: translateY(0);
  }

  &::after {
    position: absolute;
    left: 8px;
    width: 100%;
    height: 2px;
    background: $color-white1;
    content: "";
    transition: all .3s ease-in-out;
    bottom: 0;
    transform: translateY(0px);
  }
}

