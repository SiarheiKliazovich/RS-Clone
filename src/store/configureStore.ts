/* eslint-disable @typescript-eslint/ban-ts-comment */
import { applyMiddleware, createStore } from "redux";
import thunkMiddleware from "redux-thunk";

import loggerMiddleware from "./logger";
import rootReducer from "../reducers";

const configureStore = () => {
  const middlewares = [loggerMiddleware, thunkMiddleware];
  const middlewareEnhancer = applyMiddleware(...middlewares);

  const store = createStore(rootReducer, middlewareEnhancer);

  // @ts-ignore:next-line
  if (process.env.NODE_ENV !== "production" && module.hot) {
    // @ts-ignore:next-line
    module.hot.accept("./../reducers", () => store.replaceReducer(rootReducer));
  }

  return store;
};

export default configureStore;
