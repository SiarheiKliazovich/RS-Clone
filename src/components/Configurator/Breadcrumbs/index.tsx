import { useParams, useLocation } from "react-router-dom";
import { Link } from "react-router-dom";
import { useSelector } from "react-redux";
import configurator from "../../../data/configurator";
import { ILangReducer } from "../../../interfaces/langReducer";
import "./breadcrumbs.scss";
import breadcrumbsLang from "../../../data/breadcrumbs";

const Breadcrumbs = () => {
  const { lang } = useSelector((state: ILangReducer) => state.langReducer);
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const language: "ru" | "en" = lang as "ru" | "en";
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const tieList = configurator.find((c) => c.lang === lang)!;
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const breadcrumbsList = breadcrumbsLang.find((c) => c.lang === lang)!;
  const URLParams = useParams();
  const location = useLocation();

  const locationURL = location.pathname.split("/");
  console.log("locationURL", locationURL);
  const isPlains = Object.keys(URLParams).length === 0;
  const type = isPlains // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    ? locationURL[locationURL.length - 1]!
    : // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      URLParams.type!;

  const configuratorRoutes = {
    to: "/configurator",
    label: breadcrumbsList.data.config,
  };

  const baseRoutes =
    location.pathname === "/configurator"
      ? [configuratorRoutes]
      : [
          configuratorRoutes,
          { to: "/configurator/plains", label: breadcrumbsList.data.plains },
        ];

  const categoryRoutes =
    location.pathname === `/configurator/${URLParams.category}`
      ? [
          {
            to: `/configurator/${URLParams.category}`,
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            label: tieList.data.find((el) => el.type === URLParams.category)!
              .categoryName,
          },
        ]
      : [];

  const typesRoutes =
    location.pathname === `/configurator/${URLParams.category}/${type}`
      ? [
          {
            to: `/configurator/${URLParams.category}`,
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            label: tieList.data.find((el) => el.type === URLParams.category)!
              .categoryName,
          },
          {
            to: `/configurator/${URLParams.category}/${URLParams.type}`,
            label: URLParams.type,
          },
        ]
      : categoryRoutes;

  const routes = isPlains
    ? [...baseRoutes]
    : [
        { to: "/configurator", label: breadcrumbsList.data.config },
        ...typesRoutes,
      ];

  return (
    <div className="wrapper-breadcrumb">
      <div className="breadcrumb">
        <Link className="breadcrumb-link-house breadcrumb-not-active" to="/">
          <i className="fa fa-home" />
        </Link>
        {routes.map((item, key) => {
          if (routes.length - 1 === key) {
            return (
              <span
                key={item.to}
                className={
                  item.to === location.pathname
                    ? "breadcrumb-active"
                    : "breadcrumb-not-active"
                }
              >
                {item.label}
              </span>
            );
          } else {
            return (
              <Link
                key={item.to}
                className={
                  "breadcrumb-link" === location.pathname
                    ? "breadcrumb-active"
                    : "breadcrumb-not-active"
                }
                to={item.to}
              >
                {item.label}
              </Link>
            );
          }
        })}
      </div>
    </div>
  );
};

export default Breadcrumbs;
