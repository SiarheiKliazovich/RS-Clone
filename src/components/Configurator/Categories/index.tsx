import Category from "../Categories/Category";
import { FunctionComponent } from "react";
import { CategoriesType } from "../../../types/configurator";
import { ICategory } from "../../../interfaces/configurator";

const Categories: FunctionComponent<CategoriesType> = ({
  tieList,
  setType,
}: CategoriesType) => {
  return (
    <ul className="category__list">
      {tieList?.data?.map((tie: ICategory) => (
        <Category tie={tie} setType={setType} key={tie.type} />
      ))}
    </ul>
  );
};

export default Categories;
