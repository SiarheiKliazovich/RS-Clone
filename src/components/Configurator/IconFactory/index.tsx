import { FunctionComponent } from "react";
import { IconFactoryType } from "../../../types/configurator";
import { IIconMapping } from "../../../interfaces/configurator";

const IconFactory: FunctionComponent<IconFactoryType> = ({
  components,
  settings,
  type,
}: IconFactoryType) => {
  /* eslint-disable @typescript-eslint/no-non-null-assertion */
  const IconComponent = components.find(
    (component: IIconMapping) => type === component.type
  )!.component;

  return <IconComponent settings={settings} />;
};
export default IconFactory;
