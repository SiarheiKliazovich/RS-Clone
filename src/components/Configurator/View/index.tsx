import { FunctionComponent, useEffect, useState } from "react";
import classNames from "classnames";
import "./view.scss";

import colors from "../../../data/colors";
import tieWeaves from "../../../data/tieWeaves";
import { IIconSetting, IViewSetting } from "../../../interfaces/configurator";
import { ConfiguratorViewType } from "../../../types/configurator";
import getTieColor from "../../../helpers/getTieColor";
import Accordion from "../../Accordion";
import configuratorSettings from "../../../helpers/configuratorSettings";
import iconMapping from "../../../helpers/iconMapping";
import IconFactory from "../IconFactory";
import configurator from "../../../data/configurator";
import { IConfigurator } from "../../../interfaces/configurator";
import { nightTheme } from "../../../data/constants";
import { Link } from "react-router-dom";

const View: FunctionComponent<ConfiguratorViewType> = ({
  type,
  price,
  accentColor,
  isNavbarNightMode,
}: ConfiguratorViewType) => {
  const backgroundColor = nightTheme.background.element;
  const [settings, setSettings] = useState<IIconSetting>({
    weave: "",
    bgColor: "",
    colorOne: "",
    colorTwo: "",
    colorThree: "",
    colorFour: "",
    colorFive: "",
    colorSix: "",
  });

  useEffect(() => {
    setSettings({
      ...settings,
      weave: tieWeaves[0].dataSrc,
      bgColor: getTieColor(colors[0].color.join()),
    });
  }, []);

  return (
    <div className="plants__block-wrapper">
      <div className="preview-panel">
        <div className="image__wrapper">
          <IconFactory
            components={iconMapping}
            settings={settings}
            type={type}
          />
        </div>
      </div>
      <div className="edit-panel">
        <div className="edit-panel__title-wrapper">
          <h3 className="title-item">Color and Weave</h3>
        </div>
        <div className="form-wrapper">
          {configuratorSettings(type).map((setting: IViewSetting) => {
            if (setting.type === "color") {
              return (
                <Accordion title={setting.title} key={setting.title}>
                  <div className="title-form">Apply Color</div>
                  <ul className="colors-list ">
                    {colors.map((elem) => (
                      <li className="colors-item" key={elem.name}>
                        <label htmlFor="color-label">
                          <input
                            type="radio"
                            name={elem.name}
                            className="color-input"
                            onClick={() =>
                              setSettings({
                                ...settings,
                                [setting.name]: getTieColor(elem.color.join()),
                              })
                            }
                          />
                          <span
                            className="tile-picker"
                            style={{
                              backgroundColor: getTieColor(elem.color.join()),
                            }}
                          ></span>
                        </label>
                      </li>
                    ))}
                  </ul>
                </Accordion>
              );
            } else {
              return (
                <Accordion title={setting.title} key={setting.title}>
                  <div className="weave-form__wrapper">
                    <div className="title-form">Apply Weave</div>
                    <div className="weave-list-wrapper">
                      <ul className="weave-list">
                        {tieWeaves.map((tieWeave) => (
                          <li className="weave-list__item" key={tieWeave.src}>
                            <label htmlFor="" className="weave__label">
                              <input className="weave__input" type="radio" />
                              <span className="weave-span">
                                <img
                                  className={classNames("weave-img", {
                                    selected:
                                      tieWeave.dataSrc === settings.weave,
                                  })}
                                  src={tieWeave.src}
                                  alt=""
                                  onClick={() =>
                                    setSettings({
                                      ...settings,
                                      weave: tieWeave.dataSrc,
                                    })
                                  }
                                />
                              </span>
                            </label>
                          </li>
                        ))}
                      </ul>
                    </div>
                  </div>
                </Accordion>
              );
            }
          })}
        </div>
        <div className="btn-next__wrapper">
          <p className="tie-price">Price: {price}$</p>
          <Link
            to=""
            className="btn-link"
            style={{
              backgroundColor: isNavbarNightMode
                ? backgroundColor
                : accentColor,
            }}
          >
            Buy
          </Link>
        </div>
      </div>
    </div>
  );
};

export default View;
