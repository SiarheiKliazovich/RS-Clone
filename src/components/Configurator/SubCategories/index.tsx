import { FunctionComponent } from "react";
import "./subcategories.scss";
import { ISubcategory } from "../../../interfaces/configurator";
import { SubCategoriesType } from "../../../types/configurator";
import View from "../View";
import SubCategory from "./SubCategory";

const SubCategories: FunctionComponent<SubCategoriesType> = ({
  type,
  setType,
  subCategories,
}: SubCategoriesType) => (
  <div className="subcategories-wrapper">
    <h4 className="title">Tie Creators | Woven Ties | Select a Design</h4>
    {type && subCategories.length ? (
      <ul className="images-list">
        {subCategories.map((subCategory: ISubcategory) => (
          <SubCategory
            setType={setType}
            subCategory={subCategory}
            key={subCategory.id}
            type={type}
          />
        ))}
      </ul>
    ) : (
      <View type={type} />
    )}
  </div>
);

export default SubCategories;
