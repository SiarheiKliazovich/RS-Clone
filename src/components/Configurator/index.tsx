import { FunctionComponent, useState, useEffect } from "react";
import configurator from "../../data/configurator";
import Categories from "./Categories";
import SubCategories from "./SubCategories";
import { IStore } from "../../interfaces/store";
import { useSelector } from "react-redux";
import View from "./View";
import Header from "../Header";
import Profile from "../Profile";
import Footer from "../Footer";
import "./configurator.scss";

const Configurator: FunctionComponent = () => {
  const interfaceSettings = useSelector((state: IStore) => state.appInterface);
  const {
    accentColor,
    isSidebarFixed,
    isSidebarAccentMode,
    isNavbarNightMode,
    isProfileShow,
    isNightMode,
  } = interfaceSettings;

  const lang = "ru";
  const [type, setType] = useState<string>("");

  const tieList = configurator.find((c) => c.lang === lang)!;
  const typeCategory = tieList.data.find((tie) => tie.type === type)!;
  const subCategories = typeCategory?.subCategories;
  const hasSubCategories = !!typeCategory?.subCategories?.length;

  return (
    <>
      <Header
        accentColor={accentColor}
        isNavbarNightMode={isNavbarNightMode}
        isButtonVisible={false}
      />
      <Profile
        accentColor={accentColor}
        isProfileShow={isProfileShow}
        isSidebarFixed={isSidebarFixed}
        isSidebarAccentMode={isSidebarAccentMode}
        isNavbarNightMode={isNavbarNightMode}
        isNightMode={isNightMode}
      />
      <div className="configurator-wrapper">
        <div className="tie-category-banner">
          <div className="container">
            <div className="tie-category">
              {!type && !hasSubCategories ? (
                <Categories
                  tieList={tieList}
                  setType={setType}
                  type={type}
                  accentColor={accentColor}
                  isNavbarNightMode={isNavbarNightMode}
                />
              ) : type && hasSubCategories ? (
                <SubCategories
                  setType={setType}
                  type={type}
                  price={typeCategory?.price}
                  subCategories={subCategories!}
                  accentColor={accentColor}
                  isNavbarNightMode={isNavbarNightMode}
                />
              ) : (
                <View
                  type={type}
                  price={typeCategory?.price}
                  accentColor={accentColor}
                  isNavbarNightMode={isNavbarNightMode}
                />
              )}
            </div>
          </div>
        </div>
      </div>
      <Footer accentColor={accentColor} isNavbarNightMode={isNavbarNightMode} />
    </>
  );
};

export default Configurator;
