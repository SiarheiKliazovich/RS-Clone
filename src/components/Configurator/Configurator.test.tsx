import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter as Router } from "react-router-dom";
import configureStore from "redux-mock-store";
import Configurator from ".";

const mockStore = configureStore([]);

describe("Configurator", () => {
  // eslint-disable-next-line
  let store: any;

  beforeEach(() => {
    store = mockStore({
      appInterface: {
        accentColor: {
          static: "#000",
          hover: "#fff",
        },
      },
      langReducer: {
        lang: "en",
      },
    });
  });

  test("renders correctly", () => {
    const { container } = render(
      <Provider store={store}>
        <Router>
          <Configurator />
        </Router>
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });
});
