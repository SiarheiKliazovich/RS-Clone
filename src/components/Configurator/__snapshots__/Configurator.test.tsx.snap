// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Configurator renders correctly 1`] = `
<div>
  <div
    class="configurator-wrapper"
  >
    <div
      class="tie-category-banner"
    >
      <div
        class="container"
      >
        <div
          class="tie-category"
        >
          <ul
            class="category__list"
          >
            <li
              class="category__list_item"
            >
              <div
                class="category-blok"
              >
                <img
                  alt="img-ties"
                  class="category-img"
                  src="https://i.ibb.co/t2bhZVY/plains.jpg"
                />
                <a
                  class="category-name"
                  href="/configurator/plains"
                  style="background-color: rgb(0, 0, 0);"
                >
                  Plain
                </a>
              </div>
            </li>
            <li
              class="category__list_item"
            >
              <div
                class="category-blok"
              >
                <img
                  alt="img-ties"
                  class="category-img"
                  src="https://i.ibb.co/NCtWNn0/patterns.jpg"
                />
                <a
                  class="category-name"
                  href="/configurator/patterns"
                  style="background-color: rgb(0, 0, 0);"
                >
                  Patterns
                </a>
              </div>
            </li>
            <li
              class="category__list_item"
            >
              <div
                class="category-blok"
              >
                <img
                  alt="img-ties"
                  class="category-img"
                  src="https://i.ibb.co/t4RqPjL/centered.jpg"
                />
                <a
                  class="category-name"
                  href="/configurator/centred"
                  style="background-color: rgb(0, 0, 0);"
                >
                  Centred
                </a>
              </div>
            </li>
            <li
              class="category__list_item"
            >
              <div
                class="category-blok"
              >
                <img
                  alt="img-ties"
                  class="category-img"
                  src="https://i.ibb.co/2vbQXvV/simple.jpg"
                />
                <a
                  class="category-name"
                  href="/configurator/simple"
                  style="background-color: rgb(0, 0, 0);"
                >
                  Simple
                </a>
              </div>
            </li>
            <li
              class="category__list_item"
            >
              <div
                class="category-blok"
              >
                <img
                  alt="img-ties"
                  class="category-img"
                  src="https://i.ibb.co/KG8Zfhm/tartan.jpg"
                />
                <a
                  class="category-name"
                  href="/configurator/tartan"
                  style="background-color: rgb(0, 0, 0);"
                >
                  Tartan
                </a>
              </div>
            </li>
            <li
              class="category__list_item"
            >
              <div
                class="category-blok"
              >
                <img
                  alt="img-ties"
                  class="category-img"
                  src="https://i.ibb.co/0Jtdxqs/polkadot.jpg"
                />
                <a
                  class="category-name"
                  href="/configurator/polkadot"
                  style="background-color: rgb(0, 0, 0);"
                >
                  Polka Dot
                </a>
              </div>
            </li>
            <li
              class="category__list_item"
            >
              <div
                class="category-blok"
              >
                <img
                  alt="img-ties"
                  class="category-img"
                  src="https://i.ibb.co/rGPgqBx/stripedtieps.jpg"
                />
                <a
                  class="category-name"
                  href="/configurator/stripedtieps"
                  style="background-color: rgb(0, 0, 0);"
                >
                  Striped Ties
                </a>
              </div>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;
